alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs

```diff
@@ -0,0 +1,337 @@
+
+transDataHash' :: (DataHash c) -> P.DatumHash
+transDataHash' safe = (P.DatumHash (transSafeHash safe))
+
+transKeyHash :: KeyHash d c -> P.PubKeyHash
+transKeyHash (KeyHash (UnsafeHash h)) = P.PubKeyHash (fromShort h)
+
+transScriptHash :: ScriptHash i -> P.ValidatorHash
+transScriptHash (ScriptHash (UnsafeHash h)) = P.ValidatorHash (fromShort h)
+
+transSafeHash :: SafeHash c i -> BS.ByteString
+transSafeHash safe = case extractHash safe of UnsafeHash b -> fromShort b
+
+transHash :: Hash h a -> BS.ByteString
+transHash (UnsafeHash h) = fromShort h
+
+transTxId :: TxId era -> P.TxId
+transTxId (TxId safe) = P.TxId (transSafeHash safe)
+
+transStakeCred :: Credential keyrole crypto -> BS.ByteString
+transStakeCred (ScriptHashObj (ScriptHash (UnsafeHash kh))) = (fromShort kh)
+transStakeCred (KeyHashObj (KeyHash (UnsafeHash kh))) = (fromShort kh)
+
+transStakeReference :: StakeReference crypto -> Maybe P.StakingCredential
+transStakeReference (StakeRefBase cred) = Just (P.StakingHash (transStakeCred cred))
+transStakeReference (StakeRefPtr (Ptr (SlotNo slot) i1 i2)) = Just (P.StakingPtr slot i1 i2)
+transStakeReference StakeRefNull = Nothing
+
+transCred :: Credential keyrole crypto -> P.Credential
+transCred (KeyHashObj (KeyHash (UnsafeHash kh))) = P.PubKeyCredential (P.PubKeyHash (fromShort kh))
+transCred (ScriptHashObj (ScriptHash (UnsafeHash kh))) = P.ScriptCredential (P.ValidatorHash (fromShort kh))
+
+transAddr :: Addr crypto -> Maybe P.Address
+transAddr (Addr _net object stake) = Just (P.Address (transCred object) (transStakeReference stake))
+transAddr (AddrBootstrap _bootaddr) = Nothing
+
+-- ===============================
+-- Translate ValidityIntervals
+
+transVI :: ValidityInterval -> P.SlotRange
+transVI (ValidityInterval SNothing SNothing) =
+  P.Interval
+    (P.LowerBound P.NegInf True)
+    (P.UpperBound P.PosInf False)
+transVI (ValidityInterval (SJust (SlotNo i)) SNothing) =
+  P.Interval
+    (P.LowerBound (P.Finite (fromIntegral i)) True)
+    (P.UpperBound P.PosInf False)
+transVI (ValidityInterval SNothing (SJust (SlotNo i))) =
+  P.Interval
+    (P.LowerBound P.NegInf True)
+    (P.UpperBound (P.Finite (fromIntegral i)) False)
+transVI (ValidityInterval (SJust (SlotNo i)) (SJust (SlotNo j))) =
+  P.Interval
+    (P.LowerBound (P.Finite (fromIntegral i)) True)
+    (P.UpperBound (P.Finite (fromIntegral j)) False)
+
+-- ========================================
+-- translate TxIn and TxOut
+
+transTxIn' :: CC.Crypto c => TxIn c -> P.TxOutRef
+transTxIn' (TxIn txid nat) = P.TxOutRef (transTxId txid) (fromIntegral nat)
+
+-- | Given a TxIn, look it up in the UTxO. If it exists, translate it and cons the result
+--   onto the answer list. If does not exist in the UTxO, just return the answer list.
+consTranslatedTxIn ::
+  forall era.
+  ( Era era,
+    Value era ~ Mary.Value (Crypto era),
+    Core.TxOut era ~ Alonzo.TxOut era
+  ) =>
+  UTxO era ->
+  TxIn (Crypto era) ->
+  [P.TxInInfo] ->
+  [P.TxInInfo]
```

Comment:
this choice of types puts greater demands on the reader's knowledge than most alternatives

Reply:
Both now return a Maybe type.